beetsplug/discogs.py

```diff
@@ -209,8 +209,11 @@ def album_for_id(self, album_id):
         # of an input string as to avoid confusion with other metadata plugins.
         # An optional bracket can follow the integer, as this is how discogs
         # displays the release ID on its webpage.
-        match = re.search(r'(^|\[*r|discogs\.com/.+/release/)(\d+)($|\])',
+        # Issue #4080 highlighted that Discogs changed the way the URL
+        # is formed
+        match = re.search(r'(^|[*r|discogs\.com/.+/release/])(\d+)($|)',
```

Comment:
I'm not really fluent in regex, but that doesn't seem to match the description of what it's supposed to do: The regex doesn't account for the _new_ URL format, and I guess it's only working at all because the first group is a character class that basically matches anything (in particular, not only the URL). The rationale for the previous regex is given in #291, i.e. we now want to parse all of

```
just the integer: 12345
integer in brackets: [12345]
integer with prefix: r12345
discogs release url, old format: https://www.discogs.com/Larry-Heard-Sceneries-Not-Songs-Volume-One/release/15963324
discogs release url, new format: https://www.discogs.com/release/15963324-Larry-Heard-Sceneries-Not-Songs-Volume-One
```
but nothing in addition which just happens to contain a number (i.e. the first three cases are tricky, because they don't indicate that we're really dealing with a discogs ID. Thus, the corresponding regex should not match just any old URL: `yet-another-metadata-provider.org/foo/12345`).

Maybe we should really express this more explicitly, for humans to parse (untested!):
```
r'((^\[?r?\d+\]?$)|discogs\.com/release/(\d+)-|discogs\.com/[^/]+/release/(\d+))
```
instead of somehow forcing together all formats into a most compact regex.

Maybe it would also make sense to update the comment with the examples from above, and point out all of #291, #4080 and #4085.

Reply:
I changed the regex as below
```
(^|\[*r|discogs\.com.*/release/)(\d+)(|\])
```
![image](https://user-images.githubusercontent.com/1115472/135714270-9b104ba8-9c79-40c0-9cec-9e8557072d3c.png)

Reply:
This is still wrong. I'll look at it again tomorrow

Reply:
Have read the code and regex again and the regex is correct, I'll update the comment to explain it all as requested

Reply:
Updated tests

![image](https://user-images.githubusercontent.com/1115472/135733044-adace98e-c14e-4172-9b11-201504fc9d8a.png)

Reply:
This still doesn't work, since you had to remove the `$` from the last group in order to make the regex match both URLs. As a result, it now matches anything starting with an integer, for example all musicbrainz ids which happen to start with a number (i.e. it matches  `005` in `005b84a0-ecd6-39f1-b2f6-6eb48756b268`).

I still think that the best approach is to separate different things (plain id with optional brackets and two url formats) rather than code-golfing the regex. My previous suggestion

> `r'((^\[?r?\d+\]?$)|discogs\.com/release/(\d+)-|discogs\.com/[^/]+/release/(\d+))'`

 doesn't work, however, since the id would end up in different match groups. What about (again untested):
```diff
        # Discogs-IDs are simple integers. In order to avoid confusion with
        # other metadata plugins, we only look for very specific formats of the
        # input string:
        # - plain integer, optionally wrapped in brackets and prefixed by an
        #   'r', as this is how discogs displays the release ID on its webpage.
        # - legacy url format: discogs.com/<name of release>/release/<id>
        # - current url format: discogs.com/release/<id>-<name of release>
        # See #291, #4080 and #4085 for the discussions leading up to these
        # patterns.
        for pattern in [
                r'^\[?r?(?P<id>\d+)\]?$',
                r'discogs\.com/release/(?P<id>\d+)-',
                r'discogs\.com/[^/]+/release/(?P<id>\d+)',
                ]:
            match = re.search(pattern, album_id)
            if match:
                break
        if not match:
            return None
```

Reply:
Changed to the above suggestion